### PR TITLE
🎉 Source Google Sheets: Support connecting via oAuth webflow

### LINKED ISSUE #6255: Source Google Sheets: Support connecting via Oauth webflow
## Tell us about the problem you're trying to solve
With the release of Airbyte Cloud, we need to start supporting Oauth for this connector, since it's the recommended way of authenticating users into a SaaS application. 

If this connector doesn't support oauth already (i.e: doesn't accept a client_id and client_secret) then we need to update its spec to accept those parameters. I suggest that this be a oneof nested inside a top-level field called "authentication": 

```
{
  authentication: {
     type: object
     oneOf: [ 
       // api key,
       // oauth
     ] 
  }
}
```

See the connector spec reference in the docs for reference on how a oneof can be implemented.

This should be done in a backwards compatible manner i.e: users currently supplying authentication info in the config's top-level should not be impacted by this change. 

## Acceptance Criteria

1. The connector supports oauth webflow authentication with client_id/client_secret 
2. Oauth properties are annotated properly. See [this PR](https://github.com/airbytehq/airbyte/pull/6022) for an example

### PR DESCRIPTION
**Pull request by gaart**
## What
Add support for connecting via Ouath webflow

## How
Update spec.json schema to support 2 types of authorization
Update google_sheets_source.py, get_credentials method which returns credentials for a specific authorization method
Update helpers.py, add logic for choosing authorization type

## Pre-merge Checklist

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [x] Code reviews completed
- [x] Documentation updated 
    - [x] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [x] Connector version bumped like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [x] Build is successful
- [x] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [x] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [x] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

### CODE REVIEW

File: airbyte-integrations/connectors/source-google-sheets/google_sheets_source/spec.json [hunk 1]
```diff
@@ -4,18 +4,81 @@
...
+            "title": "Authenticate via Google (Oauth)",
+            "type": "object",
+            "required": [
+              "auth_type",
+              "client_id",
+              "client_secret",
+              "refresh_token"
+            ],
+            "properties": {
+              "auth_type": {
```
Comment (vitaliizazmic) [inline airbyte-integrations/connectors/source-google-sheets/google_sheets_source/spec.json:27]: I think, auth_type shouldn't be part of specification. It should be recognized in code.

File: airbyte-integrations/connectors/source-google-sheets/integration_tests/invalid_config.json [hunk 2]
```diff
@@ -1,4 +1,9 @@
...
-  "credentials_json": "{\"type\": \"service_account\",\"project_id\": \"airbyte-310409\",\"private_key_id\": \"xyz\",\"private_key\": \"-----BEGIN PRIVATE KEY-----\\n ... -----END PRIVATE KEY-----\\n\",\"client_email\": \"airbyte@airbyte-123456.iam.gserviceaccount.com\",\"client_id\": \"121512124\",\"auth_uri\": \"https://accounts.google.com/o/oauth2/auth\",\"token_uri\": \"https://oauth2.googleapis.com/token\",\"auth_provider_x509_cert_url\": \"https://www.googleapis.com/oauth2/v1/certs\",\"client_x509_cert_url\": \"https://www.googleapis.com/robot/v1/metadata/x509/airbyte%40airbyte-123456.iam.gserviceaccount.com\"}"
-}
+  "spreadsheet_id": "invalid_spreadsheet_id",
+  "credentials": {
+    "auth_type": "Client",
+    "client_id": "invalid_id",
+    "client_secret": "invalid_secret",
+    "refresh_token": "invalid_token"
+  }
+}
```
Comment (vitaliizazmic) [inline airbyte-integrations/connectors/source-google-sheets/integration_tests/invalid_config.json:9]: Did you format?

**Review by vitaliizazmic — CHANGES_REQUESTED** [review]
LGTM, but requires some improvements.

File: airbyte-integrations/connectors/source-google-sheets/google_sheets_source/spec.json [hunk 3]
```diff
@@ -4,18 +4,81 @@
     "$schema": "http://json-schema.org/draft-07/schema#",
     "title": "Stripe Source Spec",
     "type": "object",
-    "required": ["spreadsheet_id", "credentials_json"],
-    "additionalProperties": false,
+    "required": ["spreadsheet_id"],
```
[the remark this answers is not in the document]
Comment (gaart) [reply airbyte-integrations/connectors/source-google-sheets/google_sheets_source/spec.json:7]: the idea is to have a backward compatibility with the old-style config, and credentials are still required for auth through the UI

File: airbyte-integrations/connectors/source-google-sheets/google_sheets_source/spec.json [hunk 4]
```diff
@@ -4,18 +4,81 @@
...
+                  "{ \"type\": \"service_account\", \"project_id\": YOUR_PROJECT_ID, \"private_key_id\": YOUR_PRIVATE_KEY, ... }"
+                ]
+              }
+            }
+          }
+        ]
       }
     }
+  },
+  "authSpecification": {
```
[the remark this answers is not in the document]
Comment (gaart) [reply airbyte-integrations/connectors/source-google-sheets/google_sheets_source/spec.json:73]: based on #6456

File: airbyte-integrations/connectors/source-google-sheets/google_sheets_source/spec.json (hunk 1 above)
Comment (gaart) [reply airbyte-integrations/connectors/source-google-sheets/google_sheets_source/spec.json:27]: @vitaliizazmic this practice is used in the others connectors, could you point to some code with a new approach?

File: airbyte-integrations/connectors/source-google-sheets/google_sheets_source/helpers.py [hunk 5]
```diff
@@ -30,7 +31,10 @@ def get_authenticated_drive_client(credentials: Dict[str, str], scopes: List[str
 
     @staticmethod
     def get_authenticated_google_credentials(credentials: Dict[str, str], scopes: List[str] = SCOPES):
-        return service_account.Credentials.from_service_account_info(credentials, scopes=scopes)
+        if credentials.get("auth_type") == "Service":
+            return service_account.Credentials.from_service_account_info(json.loads(credentials["service_account_info"]), scopes=scopes)
+        elif credentials.pop("auth_type") == "Client":
+            return client_account.Credentials.from_authorized_user_info(info=credentials)
```
Comment (keu) [inline airbyte-integrations/connectors/source-google-sheets/google_sheets_source/helpers.py:37]: I might be wrong, but what will happen here if you first comparison will be False?
so the `auth_type` is already removed, would it always return None?

**Review by keu — CHANGES_REQUESTED** [review]
see comments

File: airbyte-integrations/connectors/source-google-sheets/google_sheets_source/helpers.py (hunk 5 above)
Comment (keu) [reply airbyte-integrations/connectors/source-google-sheets/google_sheets_source/helpers.py:37]: ok, I read it as a two `pop`s, but nevertheless can't we pop `auth_type` first and then do if/else logic?

**Review by keu — APPROVED** [review]

**Review by sherifnada — APPROVED** [review]

**Pull request merged**